File: cmd/opensave-app/frontend/src/lib/timeago.test.js
import { describe, it, expect } from 'vitest';
import { timeAgo, latestOf } from './timeago.js';

const now = Date.UTC(2026, 8, 21, 12, 0, 0);
const ago = (ms) => new Date(now - ms).toISOString();
const MIN = 60_000;
const HOUR = 60 * MIN;
const DAY = 24 * HOUR;

describe('timeAgo', () => {
  it.each([
    [ago(0), 'just now'],
    [ago(30_000), 'just now'],
    // A peer's clock a few seconds ahead of ours must not print a negative age.
    [ago(-5_000), 'just now'],
    [ago(60_000), 'a minute ago'],
    [ago(4 * MIN), '4 min ago'],
    [ago(59 * MIN), '59 min ago'],
    [ago(70 * MIN), 'an hour ago'],
    [ago(5 * HOUR), '5 hours ago'],
    [ago(26 * HOUR), 'yesterday'],
    [ago(3 * DAY), '3 days ago'],
  ])('%s → %s', (iso, want) => {
    expect(timeAgo(iso, now)).toBe(want);
  });

  it('reads the daemon\'s millisecond form and the import\'s plain RFC 3339 alike', () => {
    expect(timeAgo('2026-09-21T11:50:00.000Z', now)).toBe('10 min ago');
    expect(timeAgo('2026-09-21T11:50:00Z', now)).toBe('10 min ago');
  });

  it('gives a date beyond a week', () => {
    expect(timeAgo(ago(30 * DAY), now)).toMatch(/^on /);
  });

  it('says never for nothing, and shows garbage rather than hiding it', () => {
    expect(timeAgo('', now)).toBe('never');
    expect(timeAgo(undefined, now)).toBe('never');
    expect(timeAgo('not a time', now)).toBe('not a time');
  });
});

describe('latestOf', () => {
  it('picks the newest stamp across devices', () => {
    expect(latestOf({ deck: ago(HOUR), laptop: ago(MIN) })).toBe(ago(MIN));
  });
  it('is null with no devices synced', () => {
    expect(latestOf({})).toBeNull();
    expect(latestOf(undefined)).toBeNull();
  });
});

File: internal/cliapp/timeago_test.go
package cliapp

import (
	"testing"
	"time"
)

func TestTimeAgo(t *testing.T) {
	now := time.Date(2026, 9, 21, 12, 0, 0, 0, time.UTC)
	stamp := func(d time.Duration) string {
		return now.Add(-d).UTC().Format("2006-01-02T15:04:05.000Z")
	}
	cases := []struct {
		in   string
		want string
	}{
		{stamp(0), "just now"},
		{stamp(30 * time.Second), "just now"},
		// A peer's clock a few seconds ahead of ours must not print a
		// negative age.
		{stamp(-5 * time.Second), "just now"},
		{stamp(60 * time.Second), "a minute ago"},
		{stamp(4 * time.Minute), "4 min ago"},
		{stamp(59 * time.Minute), "59 min ago"},
		{stamp(70 * time.Minute), "an hour ago"},
		{stamp(5 * time.Hour), "5 hours ago"},
		{stamp(26 * time.Hour), "yesterday"},
		{stamp(3 * 24 * time.Hour), "3 days ago"},
		// The legacy import wrote plain RFC 3339; it must still read.
		{now.Add(-10 * time.Minute).Format(time.RFC3339), "10 min ago"},
		// Garbage is shown, not swallowed.
		{"not a time", "not a time"},
	}
	for _, c := range cases {
		if got := timeAgo(c.in, now); got != c.want {
			t.Errorf("timeAgo(%q) = %q, want %q", c.in, got, c.want)
		}
	}
	// Beyond a week the date wins; its exact text depends on the local zone,
	// so only the shape is pinned.
	if got := timeAgo(stamp(30*24*time.Hour), now); len(got) < 8 || got[:3] != "on " {
		t.Errorf("a month-old stamp printed as %q, want a date", got)
	}
}
